### ISSUE TITLE
tf.extract_image_patches gradient transpose extremely slow

### ISSUE THREAD

**Issue by ahoereth** [tensorflow/tensorflow#13017]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS 10.12.6
- **TensorFlow installed from (source or binary)**: Binary, official docker image
- **TensorFlow version (use command below)**: v1.3.0-rc2-20-g0787eee 1.3.0
- **Python version**: 3.5

### Problem
Backprop through `extract_image_patches` is extremly slow, specifically the `transpose` call. It appears to be this specific call: [tensorflow/python/ops/array_grad.py#L747](https://github.com/tensorflow/tensorflow/blob/f282bb1/tensorflow/python/ops/array_grad.py#L747)

### Logs
Timeline of the problem:

<img width="1723" alt="screen shot 2017-09-13 at 16 01 49" src="https://user-images.githubusercontent.com/3015996/30382640-f5a003b8-989f-11e7-871c-1aae1a51102b.png">

Comparable timeline when using conv2d (which, in its Eigen implementation, also [extracts image patches](https://github.com/tensorflow/tensorflow/blob/f282bb1/tensorflow/core/kernels/eigen_spatial_convolutions.h#L1063-L1064)):
<img width="1300" alt="screen shot 2017-09-13 at 16 08 45" src="https://user-images.githubusercontent.com/3015996/30382983-e898d338-98a0-11e7-9e76-1ba01e571c0d.png">

**Comment by reedwm** [comment]
@ahoereth What are the dimensions of the input tensors? Can you post code that reproduces the problem and outputs the timeline file?

**Comment by ahoereth** [comment]
You are correct, it only occurs when the input grows over a specific size.

Here a code example with some random computations:

```python
import tensorflow as tf
from tensorflow.python.client import timeline

size = 21
channels = 64

x = tf.random_uniform((100, 20))
w0 = tf.get_variable('w', (20, size * size * channels))
h = tf.reshape(tf.matmul(x, w0), (100, size, size, channels))
out = tf.extract_image_patches(images=h,
                               ksizes=(1, 5, 5, 1),
                               strides=(1, 1, 1, 1),
                               rates=(1, 1, 1, 1),
                               padding='VALID')

loss = tf.reduce_mean(out)
op = tf.train.AdamOptimizer().minimize(out)

options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
run_metadata = tf.RunMetadata()
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    sess.run(op, run_metadata=run_metadata, options=options)
    tl = timeline.Timeline(run_metadata.step_stats)
    ctf = tl.generate_chrome_trace_format()
    with open('timeline.json', 'w') as f:
        f.write(ctf)
```

Running this as above with  `size = 21` gives me the following timeline, note the time (3000ms+):
<img width="1337" alt="screen shot 2017-09-14 at 11 36 02" src="https://user-images.githubusercontent.com/3015996/30422734-1f798f8c-9941-11e7-872f-497f05fc1d80.png">

Running it with `size = 20` instead, results in this timeline (~8ms):
<img width="1306" alt="screen shot 2017-09-14 at 11 36 28" src="https://user-images.githubusercontent.com/3015996/30422785-3c849c2a-9941-11e7-8beb-55f240cd83d6.png">

We just ran this test on multiple MacBooks (mid 2015, TensorFlow 1.3) and an Ubuntu 16.04 desktop (TensorFlow 1.2.1, i7 with 4ghz, 31gb memory), all with very similar results (thanks @shoeffner).

**Comment by reedwm** [comment]
Thank you for the code sample! I can reproduce this on 1.3, but not on master, so I'm assuming its fixed and closing the issue. Feel free to reopen if you find a case where it's still slow on master.

**Comment by ahoereth** [comment]
Just ran it using the current nightly-py3 docker tag and yes, it got a lot better, but still slower by a factor of ten:

`size=21`:
<img width="1306" alt="screen shot 2017-09-14 at 19 29 23" src="https://user-images.githubusercontent.com/3015996/30444418-339d194c-9983-11e7-97c0-6008b90c91b8.png">

`size=20`:
<img width="1287" alt="screen shot 2017-09-14 at 19 30 12" src="https://user-images.githubusercontent.com/3015996/30444419-339d1320-9983-11e7-8f85-6a56eafeaa5d.png">

Will run again tomorrow to see if it changed just today, but this still seems like an extreme slow down at a very specific point (`size=22` appear to not result in performance like that anymore).

**Comment by reedwm** [comment]
You're right, it its still significantly slower when `size=21`. 

When disabling optimizations, the case where `size=20` slows down to about the speed of `size=21`, which makes me suspect certain optimizations are only applied when `size=20`. Disabling optimizations when `size=21` does not significantly affect performance.

/CC @zhangyaobit, can you comment on whether optimizations are being applied when `size=20`? A code sample with warmup batches with optmizations disabled is below:

```python
import tensorflow as tf
from tensorflow.python.client import timeline
import time

size = 21
channels = 64

with tf.device('/CPU:0'):
  x = tf.random_uniform((100, 20))
  w0 = tf.get_variable('w', (20, size * size * channels))
  h = tf.reshape(tf.matmul(x, w0), (100, size, size, channels))
  out = tf.extract_image_patches(images=h,
                                 ksizes=(1, 5, 5, 1),
                                 strides=(1, 1, 1, 1),
                                 rates=(1, 1, 1, 1),
                                 padding='VALID')

  loss = tf.reduce_mean(out)
  op = tf.train.AdamOptimizer().minimize(out)

  options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
  run_metadata = tf.RunMetadata()
  optimizer_options = tf.OptimizerOptions(opt_level=tf.OptimizerOptions.L0)
  config = tf.ConfigProto(
      graph_options=tf.GraphOptions(optimizer_options=optimizer_options))
  with tf.Session(config=config) as sess:
    sess.run(tf.global_variables_initializer())
    for _ in range(10):  # Warmup
      sess.run(op)
    start_time = time.time()
    for _ in range(50):
      sess.run(op)
    total_time = time.time() - start_time
    print('total time: {}'.format(total_time))
    sess.run(op, run_metadata=run_metadata, options=options)
    tl = timeline.Timeline(run_metadata.step_stats)
    ctf = tl.generate_chrome_trace_format()
    with open('timeline.json', 'w') as f:
      f.write(ctf)
```

**Comment by zhangyaobit** [comment]
I think you can try (1) benchmark transpose alone with size = 20 and size = 21; and (2) turn on/off optimizations individually to see which one is related (https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/core/protobuf/config.proto#L85). I know constant folding won't be applied if the size of the constant in bytes is too large (> 10M) to  prevents the size of the Graph from growing too large. But I don't see how it could be related to transpose, which I don't think it could be folded anyway.

**Comment by aselle** [comment]
@zhangyaobit, assigning to you to keep continuity of this performance problem.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by zhangyaobit** [comment]
Some performance profiling needs to be done for extract_image_patches and the transpose in the gradient op for the mentioned input sizes (preferably with two standalone benchmarks respectively for extract_image_patches's gradient op and transpose), in order to identify the slow part of the code and to suggest a potential fix.

Marking this as contribution welcome for now.

**Comment by dbbert** [comment]
Is there any progress on this issue? I'm seeing a similar dramatic slow-down when using extract_image_patches with tensors beyond a certain size.
